### ISSUE TITLE
Two-way or one-way binding in metacomponents?

### ISSUE THREAD

**Issue by Viktor286** [metarhia/metacomponents#12]
This topic is related to the core functionality of a framework and not clearly defined (at least to my understanding). Worst case scenarios might lead to essential bug behavior of a framework so i will ask this in form of a question but will provide bug-behavior examples.

What kind of binding direction with DOM metacomponents going to eventually have? two-way or one-way binding?

What behavior from "schema" object we expect in cases when someone would make DOM element update not via "schema setter" but somehow directly to DOM element via JS or CSS (for "styles" prop)?

`lib/compiler.js`
![image](https://user-images.githubusercontent.com/13506547/137840895-2c3603e7-d35f-43fa-a306-d7723dc42879.png)

It looks like currently in case of "independent-non-schema-update" the "schema" won't be in sync with DOM which would lead to some confused consequences.

In order to bring schema & the dom back in sync we would probably need to overwrite out-of-synced DOM values on the next schema-to-element "sync update" cycle? Not considering requesting this data from server each time, how would we do that schema-to-domElement "sync update" other than by re-assigning values we already have in schema "back to schema"?

In other words, in order to do "sync update" we should "get" data from "schema"... BUT it gives back data directly from "DOM" element which is out-of-sync ("independently-updated") at that moment.

So, by re-assigning values of "schema back to schema" we would assign values from DOM into the schema, which would makes domElemen-to-schema data flow direction (two-way binding). Would that be expected? Is it ok that we would loose original data from schema (until additional further request from server)?

![image](https://p88.f2.n0.cdn.getcloudapp.com/items/6quv0k9G/36b53a32-2dc0-490c-88e8-372c831d2155.gif)

**Comment by rohiievych** [comment]
We must encourage user to **not manipulate DOM directly** by providing convinient instruments in schema.
Data binding is one way (schema-to-element) by default, but you can simply catch data changes with schema events and bind them to some state.

**Comment by Viktor286** [comment]
@rohiievych Thanks for your answer! Looks like a good reasoning to me. Personally, i might prefer eventually to have a more bulletproof solution but as long as it works for you that should be sufficient.

Also it sounds like a great approach to have rehydration mechanism, that should partially address the concerns i mentioned above.

Again, thanks a lot for finding a time for detailed answer! Cheers.

**Comment by rohiievych** [comment]
Implemented particular two-way reflection via element binding.
